# **📖 Language Identification API: A Hands-on Guide**  

### **🔗 Overview**  
This notebook demonstrates how to use the **Language Identification API** to detect the language code and script code.Also we will see how we can use lamgauge identification in transalte and transliterate to auto detetct the source code and do the resepctive trnSFORMATIONS

## **Table of Contents**  

<p>1. <a href="#installation">Installation</a></p>  
<p>2. <a href="#authentication">Authentication</a></p>  
<p>3. <a href="#modes">Baisc Usage</a></p>  
<p>4. <a href="#modes">Auto Detection</a></p>  
<p>5. <a href="#error-handling">Error Handling</a></p>  
<p>6. <a href="#conclusion">Conclusion</a></p>  

## **1️⃣ Setup & Installation**  

Before you begin, ensure you have the necessary Python libraries installed. Run the following commands to install the required packages:

In [9]:
!pip install sarvamai

Defaulting to user installation because normal site-packages is not writeable
DEPRECATION: textract 1.6.5 has a non-standard dependency specifier extract-msg<=0.29.*. pip 23.3 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of textract or contact the author to suggest that they release a version with a conforming dependency specifiers. Discussion can be found at https://github.com/pypa/pip/issues/12063

[notice] A new release of pip is available: 23.2 -> 25.1.1
[notice] To update, run: /Applications/Xcode.app/Contents/Developer/usr/bin/python3 -m pip install --upgrade pip


In [10]:
from sarvamai import SarvamAI

## 2️⃣ **Authentication**


To use the API, you need an API subscription key. Follow these steps to set up your API key:

1. **Obtain your API key**: If you don’t have an API key, sign up on the [Sarvam AI Dashboard](https://dashboard.sarvam.ai/) to get one.
2. **Replace the placeholder key**: In the code below, replace "YOUR_SARVAM_AI_API_KEY" with your actual API key.

In [11]:
SARVAM_API_KEY = "YOUR_SARVAM_API_KEY"

## **3️⃣ Basic Usage**  

The API requires a single key parameter:  

✔ **`input`** – The text for which the language code and script code need to be detected.  

🚫 **Note:** If the API is unable to detect the language or script, it will return `null` for both fields.  


### **Response Parameters**  

- **`language_code`** (String) – The detected language in BCP-47 format.  Supported values:  
  - `"en-IN"` (English - India)  
  - `"en-US"` (English - United States)  
  - `"bn-IN"` (Bengali - India)  
  - `"gu-IN"` (Gujarati - India)  
  - `"hi-IN"` (Hindi - India)  
  - `"kn-IN"` (Kannada - India)  
  - `"ml-IN"` (Malayalam - India)  
  - `"mr-IN"` (Marathi - India)  
  - `"od-IN"` (Odia - India)  
  - `"pa-IN"` (Punjabi - India)  
  - `"ta-IN"` (Tamil - India)  
  - `"te-IN"` (Telugu - India)  
  - `"ur-IN"` (Urdu - India)  




- **`script_code`** (String) – The detected writing script in ISO-15924 format.Supported values:  
  - `"Latn"` → Latin (Roman script)  
  - `"Beng"` → Bengali script  
  - `"Gujr"` → Gujarati script  
  - `"Deva"` → Devanagari script  
  - `"Knda"` → Kannada script  
  - `"Mlym"` → Malayalam script  
  - `"Orya"` → Odia script  
  - `"Guru"` → Gurmukhi (Punjabi) script  
  - `"Taml"` → Tamil script  
  - `"Telu"` → Telugu script  
  - `"Arab"` → Arabic script  


In [12]:
client = SarvamAI(api_subscription_key=SARVAM_API_KEY)

In [13]:
example_text = "hey, what is your name?"

In [14]:
response = client.text.identify_language(input=example_text)
language_code = response.language_code
script_code = response.script_code

print("\n=== Detection Results ===")
print(f"Detected Language Code: {language_code}")
print(f"Detected Script Code: {script_code}\n")


=== Detection Results ===
Detected Language Code: en-IN
Detected Script Code: Latn



In [15]:
example_text = "A'in jun aatinob'aal li maare ink'a' neketaw ru."

In [16]:
response = client.text.identify_language(input=example_text)
language_code = response.language_code
script_code = response.script_code

print("\n=== Detection Results ===")
print(f"Detected Language Code: {language_code}")
print(f"Detected Script Code: {script_code}\n")


=== Detection Results ===
Detected Language Code: None
Detected Script Code: None



## **4️⃣ Auto Detection**  



To enable automatic language detection, pass `"auto"` as the `source_language_code`. The API will return the transliterated/translated text along with the detected source language code.  

🚫 **Note:** In case of detection failure, manually specify the `source_language_code` with one of the supported language codes.  


If the API is unable to detect the language, the response will include an error message:

```json
{
    "error": {
        "message": "Unable to detect the language of the input text. Please explicitly pass the `source_language_code` parameter with a supported language.",
        "code": "unprocessable_entity_error"
    }
}



### Auto Detection in Transliterate

In [17]:
response = client.text.transliterate(
    input="मुझे कल 9:30am को appointment है",
    source_language_code="auto",
    target_language_code="hi-IN",
    spoken_form=True,
)
transliterated_text = response.transliterated_text
source_language_code = response.source_language_code

print(f"✅ Transliteration Successful!\n🔤 Transliterated Text: {transliterated_text}")
print(f"🌍 Detected Source Language: {source_language_code}")

✅ Transliteration Successful!
🔤 Transliterated Text: मुझे कल सुबह साढ़े नौ बजे अपॉइंटमेंट है
🌍 Detected Source Language: hi-IN


In [18]:
response = client.text.transliterate(
    input="A'in jun aatinob'aal li maare ink'a' neketaw ru.",
    source_language_code="auto",
    target_language_code="hi-IN",
    spoken_form=True,
)
transliterated_text = response.transliterated_text
source_language_code = response.source_language_code

print(f"✅ Transliteration Successful!\n🔤 Transliterated Text: {transliterated_text}")
print(f"🌍 Detected Source Language: {source_language_code}")

UnprocessableEntityError: status_code: 422, body: {'error': {'request_id': '20250518_970f5bd8-b33d-4e34-9493-c198b9ad0952', 'message': 'Unable to detect the language of the input text, please explicitly pass the `source_language_code` parameter with a supported language.', 'code': 'unprocessable_entity_error'}}

### Auto Detection in Translate

In [19]:
response = client.text.translate(
    source_language_code="auto",
    target_language_code="bn-IN",
    speaker_gender="Male",
    mode="classic-colloquial",
    model="mayura:v1",
    input="मुझे कल 9:30am को appointment है",
)

translated_text = response.translated_text
source_language_code = response.source_language_code


print(
    f"✅ Translation Successful!\n🌍 Detected Source Language: {source_language_code}"
)
print(f"📝 Translated Text: {translated_text}")

✅ Translation Successful!
🌍 Detected Source Language: hi-IN
📝 Translated Text: আপনার জন্য কাল সকাল 9:30-এ একটা slot available আছে


## **5️⃣ Error Handling**  

You may encounter these errors while using the API:  

- **403 Forbidden** (`invalid_api_key_error`)  
  - Cause: Invalid API key.  
  - Solution: Use a valid API key from the [Sarvam AI Dashboard](https://dashboard.sarvam.ai/).  

- **429 Too Many Requests** (`insufficient_quota_error`)  
  - Cause: Exceeded API quota.  
  - Solution: Check your usage, upgrade if needed, or implement exponential backoff when retrying.  

- **500 Internal Server Error** (`internal_server_error`)  
  - Cause: Issue on our servers.  
  - Solution: Try again later. If persistent, contact support.  

- **400 Bad Request** (`invalid_request_error`)  
  - Cause: Incorrect request formatting.  
  - Solution: Verify your request structure and parameters.  

- **422 Unprocessable Entity Request** (`unprocessable_entity_error`)  
  - Cause: Unable to detect the language of the input text.
  - Solution: Explicitly pass the source_language_code parameter with a supported language.



## **6️⃣ Conclusion**  

For more details, refer to the our official documentation and we are always there to support and help you on our Discord Server:

- **Documentation**: [docs.sarvam.ai](https://docs.sarvam.ai)  
- **Community**: [Join the Discord Community](https://discord.gg/hTuVuPNF)

**Final Notes**

- Keep your API key secure.
- Use clear audio for best results.
- Explore advanced features like diarization and translation.

**Keep Building!** 🚀